Add an initial section on system constraints (#6)
abetterinternet/prio-documents: design-document.md

## System constraints

Prio has inherent constraints derived from the tradeoff between privacy
guarantees and computational complexity. These tradeoffs influence how
applications may choose to utilize services implementing the specification.

> ### Data types have limited resolution due to input-validation complexity

Reviewer: ```suggestion
### Data resolution limitations
```

Privacy comes at the cost of computational complexity. While affine-aggregatable
encodings can compute many useful statistics, they require more bandwidth and
CPU cycles to account for finite-field arithmetic during input-validation. The

> throughput of the system bounds the volume of data processed, which is

Reviewer: What is the throughput of the system? How many inputs/second can be processed?

Taking a step back: Should we have a section where we discuss this and other key performance indicators that we're trying to be mindful of? (Time between batch start and batch end, where the aggregate is available, comes to mind, for example.)

Author: I think inputs/second is a reasonable definition of throughput, which is dependent on input parameters and the resources allocated for work. I guess "rate of work" would do here too. 

Using a maybe not so great physics analogy to explain in a different frame:
* verifying input is work
* throughput is power (rate of work)
* compute-time is energy (capacity for work)
* the amount of data that can be processed is constrained by throughput and available compute-time

> Should we have a section where we discuss this and other key performance indicators that we're trying to be mindful of?

It's probably a good idea. Maybe something to consider in the requirements or a separate performance section. I imagine an implementation would also want to monitor some of these things (e.g time to completion, error counts).  This section could also be expanded for discussion.

Reviewer: > It's probably a good idea. Maybe something to consider in the requirements or a separate performance section. I imagine an implementation would also want to monitor some of these things (e.g time to completion, error counts). This section could also be expanded for discussion.

Both seem fine. Maybe just mark an OPEN ISSUE here for us to resolve later?

> proportional to the verification circuit's size and the rate of verifications

Reviewer: ```suggestion
related to the verification circuit's complexity and the rate of verifications
```

> performed by each work unit in an aggregator.

Reviewer 2: Instead of "proportional to the verification circuit's size" I would suggest "related to the circuit's complexity". The size of the proof, and the time complexity of generating and verifying it, can be reduced significantly if the circuit has a lot of redundancy.

Author: I'm interested in learning more about the optimizations at some point after overcoming my admittedly weak understanding of arithmetic circuits.

> Applications that utilize proofs with a large number of gates or a high

Reviewer: ```suggestion
Applications that utilize proofs with a large number of multiplication gates or a high
```

frequency of inputs may need to limit inputs into the system to meet bandwidth
or compute constraints. Some methods of overcoming these limitations include
choosing a better representation for the data or introducing sampling into the
data collection methodology.

> ### Aggregation services may need to fulfill near-real-time constraints

Reviewer: ```suggestion
### Aggregation utility and soft batch deadlines
```

> A near-real-time system must produce a response within a deadline. This

Reviewer: ```suggestion
A soft real-time system should produce a response within a deadline in order to be useful. This
```

Any sort of time synchronization or deadline enforcement here seems challenging, given the nature of the distributed system. I'd suggest relaxing this text a bit to say that deadlines are soft, and missing a deadline just means that the output has lesser value.

constraint may be relevant when the value of an aggregate decreases over time.

> An example of a near-real-time constraint is the expectation that a can be

Reviewer: ```suggestion
An example of a soft real-time constraint is the expectation that input data can be
```

verified and aggregated in a period equal to data collection, given some

> computational budget. Meeting these deadlines may require efficient

Reviewer 2: ```suggestion
computational budget. Meeting these deadlines will require efficient
```

implementations of the input-validation protocol. Applications might batch

> requests and utilize binary serialization to improve throughput.

Reviewer 2: What's meant by "binary serialization"?

Author: The example I had in mind was using something like Avro over JSON+base64 in cases where string parsing takes up more time than it should.

> requests and utilize binary serialization to improve throughput.

Reviewer: ```suggestion
requests or utilize more efficient serialization to improve throughput.
```

Some applications may be constrained by the time that it takes to reach a
privacy threshold defined by a minimum number of input shares. One possible
solution is to increase the reporting period so more samples can be collected
before the period is over, although this may need to be balanced with the need

> to produce a response within a deadline.

Reviewer: ```suggestion
to produce a response within a soft deadline.
```

### Aggregation services may need to fulfill data integrity constraints

Data integrity concerns the accuracy and correctness of the outputs in the
system. The integrity of the output can be influenced by an incomplete round of
aggregation caused by network partitions, or by bad actors attempting to cause

> inaccuracies in the aggregates. An example data integrity constraint may be that

Reviewer 2: ```suggestion
inaccuracies in the aggregates. An example data integrity constraint is that
```
Definitely an important constraint :)

> every share must be processed at most once by all aggregators. Data integrity

Reviewer 2: ```suggestion
every share must be processed exactly once by all aggregators. Data integrity
```

constraints may be at odds with the threat model if meeting the constraints
requires replaying data.

Invalid shares are expected when operating an aggregator and do not necessarily
constitute failures in the protocol. However, a large number of invalid

> or identical inputs may be concerning participants in the protocol.

Reviewer 2: Invalid shares are always a problem if they go undetected. It sounds like this paragraph is more about DoS mitigation? I.e., regardless of their validity, the aggregators need to be able to detect if they're being flooded by inputs?

Author: I mean for this paragraph to distinguish between normal and unusual. In my day-to-day work, I've trawled through garbage coming through public ingestion endpoint (a lot of fuzzers and malformed data). I think some error is tolerable and normal. A large number of invalid inputs could be attributed to misconfigured application, and not necessarily an explicit denial of service. I did have DoS in mind too; I'll be more specific.

Certain classes of errors do not exist in the input-validation protocol
considered in this document. For example, errors from packet loss when clients
make requests directly to aggregators are not relevant when the leader proxies

> requests and controls the clock cycle for signaling the end of the aggregation

Reviewer: ```suggestion
requests and controls the schedule for signaling the end of the aggregation
```

> rounds.

Reviewer 2: This is good stuff. I feel like we're definitely gonna expand this paragraph as we implement this :)
